policy/serialization_test.go

```diff
@@ -26,74 +28,66 @@ func assertEquals(t *testing.T, actual interface{}, expected interface{}) {
     if err == nil {
-	t.Errorf("Expected operator %v to produce an InvalidOperatorError", invalid.OOperator)
-    } else {
-	jsonError, _ := err.(*json.MarshalerError)
-	switch jsonError.Err.(type) {
-	case InvalidOperatorError:
-	default:
-	    t.Errorf("Expected operator %v to produce an InvalidOperatorError but got: %T", invalid.OOperator, err)
-	}
+	t.Errorf("Expected operator %v to produce an error", invalid.OOperator)
+    }
+    jsonError, _ := err.(*json.MarshalerError)
+    assertEquals(t, jsonError.Err.Error(), "Invalid operator: '15'. Valid values are AND, OR, NONE")
+}
+
+func TestDeserializeMissingOpertor(t *testing.T) {
+    pl := PolicyLine{}
+    err := json.Unmarshal([]byte("{}"), &pl)
+    if err == nil {
+	t.Errorf("Expected deserialization to fail when operator is missing")
     }
+    assertEquals(t, err.Error(), "Invalid operator: ''. Valid values are AND, OR, NONE")
 }
 
-var jsonRuleSet = "{\"op\":\"OR\"," +
-		    "\"left\":{\"op\":\"AND\"," +
-			"\"left\":{\"op\":\"NONE\",\"rule\":{\"layer\":\"l1\",\"layer_type\":\"t1\",\"pattern\":\"p1\"}}," +
-			"\"right\":{\"op\":\"NONE\",\"rule\":{\"layer\":\"l2\",\"layer_type\":\"t2\",\"pattern\":\"p2\"}}}," +
-		    "\"right\":{\"op\":\"NONE\",\"rule\":{\"layer\":\"l3\",\"layer_type\":\"t3\",\"pattern\":\"p3\"}}}"
+var c1 = Credential{ Name: "n1", Value: "v1" }
+var forever =  Duration{ time.Date(0, 1, 1, 0,0,0,0, time.UTC), time.Date(3000, 1,1, 0,0,0,0, time.UTC) }
+var everywhere = Location { "everywhere" }
 
-var rs1 = RuleSet{OOperator: NONE, RRule: &Rule{Layer: "l1", LType: "t1", Pattern: "p1"}}
-var rs2 = RuleSet{OOperator: NONE, RRule: &Rule{Layer: "l2", LType: "t2", Pattern: "p2"}}
-var rs3 = RuleSet{OOperator: NONE, RRule: &Rule{Layer: "l3", LType: "t3", Pattern: "p3"}}
-var rsAnd = RuleSet{OOperator: AND, LArg: &rs1, RArg: &rs2}
-var rsOr = RuleSet{OOperator: OR, LArg: &rsAnd, RArg: &rs3}
+var tcp80Name = makeIPRule("10.0.0.1").And(makeTCPRule("80")).And(makeServiceRule("/home"))
+var tcp443Name = makeIPRule("10.0.0.1").And(makeTCPRule("443")).And(makeServiceRule("/home"))
 
-func TestRuleSetSerialization(t *testing.T) {
-    serialized, err := json.Marshal(&rsOr)
+var tcp80Resource = Resource{ Name: tcp80Name, IdentifiedBy: &c1 }
+var tcp443Resource = Resource{ Name: tcp443Name, IdentifiedBy: &c1 }
+
+// This is tautological, but we're testing the policy line matcher not the deeper policy logic
+var tcp80Policy = makePolicy(tcp80Resource, &tcp80Resource, nil,  forever, everywhere)
+var tcp443Policy = makePolicy(tcp443Resource, &tcp443Resource, nil,  forever, everywhere)
+
+var tcp80PolicyLine = PolicyLine{ OOperator: NONE, PPolicy: tcp80Policy }
+var tcp443PolicyLine = PolicyLine{ OOperator: NONE, PPolicy: tcp443Policy }
+var tcp80or443PolicyLine = PolicyLine{ OOperator: OR, LArg: &tcp80PolicyLine, RArg: &tcp443PolicyLine }
+
+func TestPolicySerialization(t *testing.T) {
+    var jsonPolicy, err = ioutil.ReadFile("./test_policy.json")
```

Comment:
how was the file generated and are there instructions on how to generate it so that if we change the structure we're not generating it by hand.

Reply:
I did it by serializing the object:
```go
str, err := json.MarshalIndent(&bundle, "", "  ")
```
I manually compacted the json a bit to make it easier to read and understand. Just for the review purposes, as we want to clearly see how defining policies would look like.